src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -876,5 +879,68 @@ private static int LocateLastFoundChar(ulong match)
         {
             return 3 - (BitOperations.LeadingZeroCount(match) >> 4);
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref char Add(ref char source, nint elementOffset)
+            => ref Unsafe.Add(ref source, (IntPtr)elementOffset);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe Vector<ushort> LoadVector(ref char start, nint offset)
+            => Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref start, (IntPtr)offset)));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe Vector128<ushort> LoadVector128(ref char start, nint offset)
+            => Unsafe.ReadUnaligned<Vector128<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref start, (IntPtr)offset)));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe Vector256<ushort> LoadVector256(ref char start, nint offset)
+            => Unsafe.ReadUnaligned<Vector256<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref start, (IntPtr)offset)));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe UIntPtr LoadUIntPtr(ref char start, nint offset)
+            => Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref start, (IntPtr)offset)));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe nint GetCharVectorSpanLength(nint offset, nint length)
+            => ((length - offset) & ~(Vector<ushort>.Count - 1));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe nint GetCharVector128SpanLength(nint offset, nint length)
+            => ((length - offset) & ~(Vector128<ushort>.Count - 1));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static nint GetCharVector256SpanLength(nint offset, nint length)
+            => ((length - offset) & ~(Vector256<ushort>.Count - 1));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe nint UnalignedCountVector(ref char searchSpace)
+        {
+            const int elementsPerByte = sizeof(ushort) / sizeof(byte);
+            // Figure out how many characters to read sequentially until we are vector aligned
+            // This is equivalent to:
+            //         unaligned = ((int)pCh % Unsafe.SizeOf<Vector<ushort>>()) / elementsPerByte
+            //         length = (Vector<ushort>.Count - unaligned) % Vector<ushort>.Count
+            int unaligned = ((int)Unsafe.AsPointer(ref searchSpace) & (Unsafe.SizeOf<Vector<ushort>>() - 1)) / elementsPerByte;
+            return ((Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1));
```

Comment:
More optimized version (4 instructions instead of 6):

```cs
return (-(int)Unsafe.AsPointer(ref searchSpace) / elementsPerByte) & (Vector<ushort>.Count - 1);
```

Additionally, since this method requires _searchSpace_ to be pinned (otherwise immediately after the call to `AsPointer` it could be relocated by the GC), consider taking the parameter as a `char*` directly rather than as a `ref char`.

(This same comment applies to the methods immediately below, though the bit twiddling may need to change.)

Reply:
Not really fussed worried about relocation; it acts as a hint (hence using read unaligned for all the reads). If GC occurs the misalignment cost is likely dwarfed by the GC cost itself?

Will add comment to this effect though.

Reply:
That's a good point. Since GC could be happening on another thread I don't know what the true cost would be. @jkotas or @maoni0 would be able to offer better advice.

Reply:
> If GC occurs the misalignment cost is likely dwarfed by the GC cost itself?

Correct.

Reply:
Addressed feeback